### ISSUE TITLE
How to ignore optional chaining and nullish coalescing branches

### ISSUE THREAD

**Issue by Shyiy** [istanbuljs/istanbuljs#526]
After the adoption of TypeScript 3.7 (https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html) and refactoring of code I use in many places the new `?.` and `??` operators.

I obviously try to cover each case via tests but in some cases I would like to ignore some specific branch resulting from the new operators. I tried several different ways to add "istanbul ignore ..." between symbols but nothing seems to work. Is it possible to ignore only some of the branches without ignoring the full line or rewriting it with the older, more verbose syntax? Or (worst) ignore a line only for branch coverage?

Some examples:
```
const a = b ?? c; // Ignore uncovered "c" case
```
```
e?.f?.g(); // Ignore uncovered cases where "e" or "f" are undefined
```

**Comment by coreyfarrell** [comment]
Can you comment with an example of output from typescript for the following:
```js
/* before const */
const a = /* before b */ b /* before ?? */ ?? /* after ?? */ c;
```

Is it valid to add comments before/after `?.` in optional chaining?  If so please show the result of:
```js
/* before optional chaining */
e /* after e */ ?. /* before f */ f /* after f */ ?. /* before g */ g();
```

@bcoe @JaKXz you have more experience than I with TS, any thoughts on this issue / possible work arounds?

**Comment by Shyiy** [comment]
For the `?.` operator it's legal to put comments before or after it but that doesn't seem to help.

You can play around and easily see the output code in this playground:
https://www.typescriptlang.org/play/#code/PQKgBARgpgZg9gJymAxnAdgZwC5hMAKDS1wEMwBeMUSWRZCPYSa8aeJMAfi6e95qkY2KAn58UAbgIAbKLijSa7emDgAHbAEsMpGagAWpLehMBzJgSiDhosMnxcAdMrqcYTGDZFiPjl2xuyBb4YGYAFACUkkA

For this particular example:
```
/* before const */
a = /* before b */ b /* before ?? */ ?? /* after ?? */ c;
/* before optional chaining */
e/* after e */?./* before f */f/* after f */?./* before g */g();
```
becomes:
```
"use strict";
var _a, _b;
/* before const */
a = /* before b */ (b /* before ?? */ !== null && b /* before ?? */ !== void 0 ? b /* before ?? */ : c);
/* before optional chaining */
(_b = (_a = e /* after e */) === null || _a === void 0 ? void 0 : _a.f /* after f */) === null || _b === void 0 ? void 0 : _b.g();
```

**Comment by coreyfarrell** [comment]
I don't think istanbuljs is able to support coverage hints in the code generated by TypeScript for these new JS features.  The only suggestion I can provide right now is:
```js
/* istanbul ignore next */
a = b ?? c;
/* istanbul ignore next */
e?.f?.g();
```

Not optimal but as long as these statements are inside a function or a conditional branch you should be able to see if they are not executed at all.  For example:
```js
function fnWithNullish(b, c) {
  /* istanbul ignore next */
  return b ?? c;
}
```

In this example istanbul will report that the function `fnWithNullish` was never executed but will not report on the `return b ?? c` statement.

**Comment by Shyiy** [comment]
I see, too bad for now...
Would it be possible maybe in the future as a partial improvement for this or other similar situations then to have some new keywords in the istanbul ignore command to specify which type of coverage needs to be ignored instead of everything?

Something like:
```
// istanbul ignore branch next
```

That way the code wouldn't need to be artificially refactored in order to provide at least line/statement coverage.

**Comment by coreyfarrell** [comment]
Any change to `istanbul ignore` hints is high risk and complex.  I don't have time to go down this rabbit hole, I'm not sure istanbuljs can produce better results from the code typescript generates for these two features.  TypeScript output is often unfriendly to coverage, this is a common problem.  The more TypeScript transpiles the bigger the problem becomes.

The only thought I have is if typescript has an option that can disable optional chaining, then you might be able to use babel to handle this.  Then babel would be able to run `babel-plugin-istanbul` before the optional chaining / nullish coalescing plugins.  Something to understand both of these syntaxes are new features and it will take some time before tooling support matures.

**Comment by Shyiy** [comment]
No problem, I understand. Thanks for the confirmation!

**Comment by coreyfarrell** [comment]
Note that istanbul-lib-instrument already supports ignoring nullish coalescing branches for non-transpiled code:
```js
const a = b ?? /* istanbul ignore next */ c;
```

It's possible that `ts-node/register` might help here.  The idea is that `istanbul-lib-instrument` can add coverage counters to the TypeScript source then `ts-node/register` can run on the instrumented code.

Side effect of this istanbul-lib-instrument does not collect branch coverage for optional chaining operators.  I'm looking into adding support for this but it's difficult.

**Comment by pbn04001** [comment]
@coreyfarrell Can you provide an example of how to set this up globally to ignore nullish coalescing for the entire project?

**Comment by coreyfarrell** [comment]
@pbn04001 this is not possible and no such option is planned.   Nullish coalescing is an expression branch and if you don't exercise all of them you don't have 100% coverage.  Nullish coalescing works the same as `&&` / `||` for instrumentation.

**Comment by coreyfarrell** [comment]
@dhon istanbul does not support optional chaining at all.  What you are seeing is the result of something else transpiling `?.` into conditional code, istanbul is then seeing that conditional code and producing coverage of that.

Based on the coverage you are posting you are not testing the situation where `this.route == null`.  `this.route && this.route.snapshot` only creates a branch for `!!this.route` (this.route not falsy).  The code generated by your transpiler for `this.route?.snapshot` is (likely) creating a branch for `this.route` nullish and another for not nullish.  It's possible that it creates 2 branches for nullish - one for `this.route === null` and another for `this.route === undefined`.  From what I'm seeing you have source-maps correctly enabled and your transpiler is associating `?.` from source to generated code.

**Comment by dhon** [comment]
@coreyfarrell Thanks, yeah that's what I'm doing in the spec file. I was just thinking it would be nice if optional chaining could work the same as && chaining in regards to branch coverage, either by default or a setting, but as you mentioned it's not supported.

**Comment by anmolsinha1793** [comment]
Is there an update on the above issue?

**Comment by PRossetti** [comment]
I'm having the same issue here. Code like this would be mark as 100% covered by jest

if (body && body.foo && body.foo.faa) { ... }

But this equivalent code won't. The ?. won't be marked as covered.
`if (body?.foo?.faa) { ... }`

Having an easy elegant way of getting pass by this behaviour would be very much appreciated.

**Comment by joshuambg** [comment]
Please do not suggest this solution from a previously **incorrectly** closed issue.
> Hi there, it worked for me:
> 
> * Upgrade your node version to at least Node 14
> * Update your tsconfig.js `target` attribute to `es2020`
>
> https://github.com/istanbuljs/istanbuljs/issues/516

This issue is NOT fixed and persists. Updating  tsconfig.js `target`is not an acceptable solution. Projects are not always able to target `es2020`. 

This issue persists and needs a real solution added to istanbul core.

**Comment by jiangzongyu** [comment]
any news?

**Comment by Akash280899** [comment]
> @coreyfarrell Can you provide an example of how to set this up globally to ignore nullish coalescing for the entire project?

Yes. It would be great if we have global ignore case for this. Because if we have so many files having this optional chaining and everywhere I want to go to put istanbul ignore, also that's not feasible. Searched in many places not found anything.

**Comment by EuginFrazz** [comment]
Any effective solution for this?

**Comment by willbui256resi** [comment]
Still no fix?

**Comment by aczekajski** [comment]
This is still a problem...
I have a code like this:
```
a.b?.forEach( ... );
```
and when I put `/* istanbul ignore next */` above this line, the entire callback function of that `forEach` is ignored as well, which I don't want. The only solution is to split it to separate statements but doing so just for the sake of code coverage feels off...

**Comment by aczekajski** [comment]
This thread being 5 years old and still having no solution other than splitting such expressions into multiple separate assignments is ridiculous...
